shared/modules/bridge-utils/balance.ts

```diff
@@ -0,0 +1,64 @@
+import { Web3Provider } from '@ethersproject/providers';
+import { Hex } from '@metamask/utils';
+import { zeroAddress } from 'ethereumjs-util';
+import { getAddress } from 'ethers/lib/utils';
+import { HttpProvider } from '../../../types/global';
+import { fetchTokenBalance } from '../../lib/token-util';
+import { Numeric } from '../Numeric';
+
+export const calcLatestSrcBalance = async (
+  provider: HttpProvider,
+  selectedAddress: string,
+  tokenAddress: string,
+  chainId: Hex,
+): Promise<Numeric | undefined> => {
+  if (tokenAddress && chainId) {
+    if (tokenAddress === zeroAddress()) {
+      const ethersProvider = new Web3Provider(provider);
+      return Numeric.from(
+        (
+          await ethersProvider.getBalance(getAddress(selectedAddress))
+        ).toString(),
+        10,
+      );
+    }
+    return Numeric.from(
+      (
+        await fetchTokenBalance(tokenAddress, selectedAddress, provider)
+      ).toString(),
+      10,
+    );
+  }
+  return undefined;
+};
+
+export const hasSufficientBalance = async (
```

Comment:
The logic in this function is a little odd to me. Can't tokenAddress be both ZeroAddress or a TokenAddress?

If it's a ZeroAddress then we're checking that `balance > fromTokenAmount` and `balance > 0`

what is the point of `isNativeBalanceGreaterThanZero`?

shouldn't we have logic that more closely represents:

Case1 ERC20:
tokenBalance >= fromTokenAmount && nativeTokenBalance > gasCost

Case2 Native:
nativeTokenBalance > fromTokenAmount + gasCost

Reply:
Nice catch! The purpose of `isNativeBalanceGreaterThanZero` was to check if the user has any native tokens to cover gas when bridging an ERC20 src token. My assumption was that the backend simulation would fail if we don't pass in  `insufficientBal=true` when the user has enough ERC20 but insufficient enough native assets

Just checked this and the bridge-api can handle this scenario (we still get quotes) so `isNativeBalanceGreaterThanZero` can be removed

Reply:
`hasSufficientBalance` currently only runs prior to fetching quotes in order to set `insufficientBal`. We'll need to add a separate check to disable/enable submission based on `gasCost` after the total network fees are calculated

I'll update this to have this logic:

Case1 ERC20:
tokenBalance >= fromTokenAmount

Case2 Native:
nativeTokenBalance >= fromTokenAmount